lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -82,46 +100,60 @@ withNetworkLayer
+    , currentNodeTip
     , currentNodeEra = undefined
     , currentProtocolParameters = undefined
     , currentNodeProtocolParameters = undefined
     , currentSlottingParameters = undefined
-    , watchNodeTip = watchNodeTip
+    , watchNodeTip
     , postTx = undefined
     , stakeDistribution = undefined
     , getCachedRewardAccountBalance = undefined
     , fetchRewardAccountBalances = undefined
     , timeInterpreter = undefined
     , syncProgress = undefined
     }
-
   where
+    currentNodeTip :: IO BlockHeader
+    currentNodeTip = runExceptT fetchLatestBlockHeader >>= \case
+        -- TODO: use cached value while retrying
+        Left err -> throwIO (BlockfrostException err)
+        Right header -> pure header
 
     watchNodeTip :: (BlockHeader -> IO ()) -> IO ()
     watchNodeTip callback = link =<< async (pollNodeTip callback)
+      where
+        pollNodeTip :: (BlockHeader -> IO ()) -> IO ()
+        pollNodeTip cb = forever $ do
+            runExceptT fetchLatestBlockHeader >>= \case
+                Left err -> throwIO (BlockfrostException err)
+                Right header ->
+                    bracketTracer (MsgWatcherUpdate header >$< tr) $ cb header
+            threadDelay 2_000_000
 
-    pollNodeTip :: (BlockHeader -> IO ()) -> IO ()
-    pollNodeTip callback = forever $ do
-        runExceptT fetchLatestBlockHeader >>= \case
-            Left err -> traceWith tr (MsgBlockfrostClientError err)
-            Right header ->
-                bracketTracer (MsgWatcherUpdate header >$< tr) $ callback header
-        threadDelay 2_000_000
+    fetchLatestBlockHeader :: ExceptT BlockfrostError IO BlockHeader
+    fetchLatestBlockHeader =
+        runBlockfrost BF.getLatestBlock >>= blockToBlockHeader
 
     runBlockfrost :: BF.BlockfrostClientT IO a -> ExceptT BlockfrostError IO a
     runBlockfrost =
         ExceptT . (first ClientError <$>) . BF.runBlockfrostClientT project
 
-    fetchLatestBlockHeader :: ExceptT BlockfrostError IO BlockHeader
-    fetchLatestBlockHeader = do
-        block@BF.Block{..} <- runBlockfrost BF.getLatestBlock
-        slotNo <- case _blockSlot of
-            Just (BF.Slot s) -> pure $ SlotNo $ fromIntegral s
-            Nothing -> throwError $ NoSlotError block
-        pure BlockHeader
-            { slotNo
-            , blockHeight = undefined -- Quantity "block" Word32
-            , headerHash = undefined -- !(Hash "BlockHeader")
-            , parentHeaderHash = undefined -- !(Maybe (Hash "BlockHeader"))
-            }
+blockToBlockHeader ::
+    forall m. MonadError BlockfrostError m => BF.Block -> m BlockHeader
+blockToBlockHeader block@BF.Block{..} = do
+    slotNo <- case _blockSlot of
+        Just s -> pure $ SlotNo $ fromIntegral $ BF.unSlot s
+        Nothing -> throwError $ NoSlotError block
+    blockHeight <- case _blockHeight of
+        Just height -> pure $ Quantity $ fromIntegral height
+        Nothing -> throwError $ NoBlockHeight block
```

Comment:
🤔 I think the `Nothing` case indicates the genesis block, but I'm not entirely certain. I recently learned that in the Byron era, there exist Epoch Boundary Blocks (EBBs) which are weird. Could you check (or google) what value Blockfrost returns for the genesis block and these special blocks?

The Cardano mainnet genesis is:
https://explorer.cardano.org/en/block?id=5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb

The block after genesis is one of those fabled EEBs:
https://explorer.cardano.org/en/block?id=89d9b5a5b8ddc8d7e5a6795e9774d97faf1efea59b2caf7eaf9f8c5b32059df4